torch/distributed/_sharded_tensor/api.py

```diff
@@ -750,3 +796,61 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
                           device=local_device, )
     else:
         raise ValueError(f'Unsupported create_op: {tensor_init_params.create_op}')
+
+
+class PartialTensor(object):
+    """
+    PartialTensor is an abstraction to represent Tensors that need
+    aggregation across multiple devices and multiple processes.
+
+    PartialTensor is initialized in an SPMD like fashion where each rank
+    initializes the PartialTensor. The PartialTensor object on each rank
+    then only stores the local partial shard and provides global metadata
+    for all the shards.
+
+    PartialTensor doesn't provide any Tensor like operations but is a
+    wrapper providing the Tensor representing the local partial shard
+    and the global metadata.
+
+    Using these, users can apply custom distributed sharded computations
+    on top of this primitive. The local shards are all initialized using
+    the create_op specified by tensor_init_params.create_op, e.g.,
+    torch.ones, or torch.empty.
+
+    Args:
+        local_partial_results (List(Tensor)):
+            Partial results are stored across ranks and in a form of a list.
+    """
+
+    def __init__(self, local_partial_results):
+        self.local_partial_results = local_partial_results
+
+    def __post_init__(self):
+        if not isinstance(self.local_partial_results, Iterable):
+            raise ValueError('local_partial_results needs to be a Iterable.')
+        for tensor in self.local_partial_results:
+            if not isinstance(tensor, torch.Tensor):
+                raise ValueError('Element in local_partial_results needs to be a Tensor.')
+
+    def reshard(self, _resharding_spec: ReshardingSpec) -> torch.Tensor:
+        """
+        Consolidate all partial tensors across multiple ranks based on resharding
+        spec.
+        """
+        local_result = torch.empty_like(self.local_partial_results[0])
+        return reduce_scatter(local_result, self.local_partial_results)
+
+
+class ModuleResharder(torch.nn.Module):
```

Comment:
make this private too since it is for internal use.

Reply:
Ah this is indeed a much better solution since it keeps the the original module intact as well. @fduwjj We should try out this hook to see if it fits our use case.

Reply:
Sure. Let me look into it.

Reply:
Can we use this hook for the prod use? The wiki says "This adds global state to the nn.module module and it is only intended for debugging/profiling purposes.". We probably need to go with our previous proposal..

Reply:
Done. The hook way works.